File: import.py
# @copyright 2018 City of Bloomington, Indiana
# @license http://www.gnu.org/licenses/gpl.txt GNU/GPL, see LICENSE.txt
#
# This file is part of ckan_google_importer.
#
# ckan_google_importer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ckan_google_importer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ckan_google_importer.  If not, see <https://www.gnu.org/licenses/>.
import argparse
import errno
import os
import pycurl
import sys
import yaml

parser = argparse.ArgumentParser()
parser.add_argument("configfile", help="path to configuration YML file")
args = parser.parse_args()

if len(sys.argv)==1:
    parser.print_help(sys.stderr)
    sys.exit(1)

with open(args.configfile) as stream:
    try:
        doc = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)
        sys.exit(1)

tmp_dir  = doc['tmp_dir' ]
log_file = doc['log_file']
ckan_url = doc['ckan']['url'    ] + '/api/3/action/resource_update'
api_key  = doc['ckan']['api_key']

if not os.path.isdir(tmp_dir):
    print(tmp_dir + ' does not exist')
    sys.exit(1)

download = pycurl.Curl();
upload   = pycurl.Curl();
download.setopt(pycurl.FOLLOWLOCATION, True)
upload.setopt(pycurl.URL, ckan_url)
upload.setopt(pycurl.HTTPHEADER, ['Authorization: ' + api_key])

for s in doc['spreadsheets']:
    resource_id    = s['resource_id'   ]
    spreadsheet_id = s['spreadsheet_id']
    sheet_id       = s['sheet_id'      ]
    google_url     = 'https://docs.google.com/spreadsheets/d/' + spreadsheet_id + '/export?format=csv&gid=' + sheet_id
    csv_file       = tmp_dir + '/' + resource_id + '.csv'

    with open(csv_file, 'wb') as f:
        download.setopt(pycurl.URL,       google_url)
        download.setopt(pycurl.WRITEDATA, f)
        download.perform()

        upload.setopt(pycurl.HTTPPOST, [
            ('id',     resource_id),
            ('upload', (
                pycurl.FORM_FILE,        csv_file,
                pycurl.FORM_CONTENTTYPE, 'text/csv'
            ))
        ])
        upload.perform()

download.close()
upload.close()

File: import.php
<?php
/**
 * @copyright 2019 City of Bloomington, Indiana
 * @license https://www.gnu.org/licenses/agpl.txt GNU/AGPL, see LICENSE
 */
declare (strict_types=1);
include './SkidderService.php';

$opts = getopt('', ['help', 'config:']);
if (isset($opts['help'])) {
    echo <<<EOT
DESCRIPTION
    This script downloads Google Spreadsheets as CSV files, then uploads
    them to resources in a CKAN site.

OPTIONS
    Generic Program Information
        --help   Output a usage message and exit
        --config Path to the configuration file

EOT;
exit();
}

if (     isset($opts['config'])
    &&  !empty($opts['config'])
    && is_file($opts['config'])) {

    $config = json_decode(file_get_contents($opts['config']));
}

if (!isset($config) || !$config) {
    echo "import: you must specify the path to a valid json config file\n";
    exit(1);
}

if (!is_dir($config->tmp_dir)) {
    echo "import: you must configure a tmp directory to store downloaded files\n";
    exit(1);
}

$tmp_dir  = $config->tmp_dir;
$ckan_url = $config->ckan->url . '/api/3/action/resource_update';
$api_key  = $config->ckan->api_key;

$download = curl_init();
$upload   = curl_init();
curl_setopt($download, CURLOPT_FOLLOWLOCATION, true);
curl_setopt_array($upload, [
    CURLOPT_URL            => $ckan_url,
    CURLOPT_POST           => true,
    CURLOPT_HEADER         => false,
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_SSL_VERIFYPEER => false,
    CURLOPT_HTTPHEADER     => ["Authorization: $api_key"]
]);

foreach ($config->spreadsheets as $s) {
    $name           = $s->name;
    $resource_id    = $s->resource_id;
    $spreadsheet_id = $s->spreadsheet_id;
    $sheet_id       = $s->sheet_id;
    $google_url     = "https://docs.google.com/spreadsheets/d/$spreadsheet_id/export?format=csv&gid=$sheet_id";
    $csv_file       = "$tmp_dir/$name.csv";

    $fp = fopen($csv_file, 'w');
    curl_setopt_array($download, [CURLOPT_URL  => $google_url,
                                  CURLOPT_FILE => $fp        ]);
    $success = curl_exec($download);
    if (!$success) {
        SkidderService::log($config->skidder->url, $config->skidder->application_id, [
            'script'  => $google_url,
            'type'    => 'Curl error',
            'message' => curl_error($download)
        ]);
    }
    fclose($fp);

    curl_setopt($upload, CURLOPT_POSTFIELDS, [
        'id'     => $s->resource_id,
        'upload' => new \CURLFile($csv_file, 'text/csv', 'data.csv')
    ]);
    $response = curl_exec($upload);
    if (!$response) {
        SkidderService::log($config->skidder->url, $config->skidder->application_id, [
            'script'  => $ckan_url,
            'type'    => 'Curl error',
            'message' => "$csv_file\n{$s->resource_id}\n".curl_error($upload)
        ]);
    }
    else {
        $json = json_decode($response);
        if (!$json) {
            SkidderService::log($config->skidder->url, $config->skidder->application_id, [
                'script'  => $ckan_url,
                'type'    => 'CKAN invalid response',
                'message' => "$csv_file\n{$s->resource_id}\n$response"
            ]);
        }

        if (!$json->success) {
            SkidderService::log($config->skidder->url, $config->skidder->application_id, [
                'script'  => $ckan_url,
                'type'    => 'CKAN error: '.$json->error->__type,
                'message' => "$csv_file\n{$s->resource_id}\n".$json->error->message
            ]);
        }
    }
}
curl_close($download);
curl_close($upload);
